enhancements/scheduling/prevent-user-workloads-from-being-scheduled-on-control-plane-nodes.md

```diff
@@ -0,0 +1,398 @@
+
+```
+oc label <NAMESPACE_NAME> openshift.io/control-plane=true
+```
+
+5. Pod Scheduling Attempt:
+
+- Users/Developers proceed to deploy or update pods, embedding scheduling preferences, such as necessary tolerations. Example of adding a relevant toleration:
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: my-pod
+  namespace: <NAMESPACE_NAME>
+  labels:
+    app: my-app
+spec:
+  containers:
+  - name: my-container
+    image: nginx
+    ports:
+    - containerPort: 80
+  nodeName: ip-XX-X-XX-XXX.ec2.internal
+  tolerations:
+  - key: "node-role.kubernetes.io/control-plane"
+    operator: "Exists"
+    effect: "NoExecute"
+```
+
+- These requests are then submitted to the Kubernetes API server for processing.
+
+6. Admission Policy Enforcement:
+
+- Upon receiving pod deployment or update requests, the Kubernetes API server invokes the ValidatingAdmissionPolicy. It evaluates the pod's tolerations against the policy's defined CEL expressions, determining the request's compatibility with established configured guidelines.
+
+7. Policy Decision Feedback:
+
+- Should a pod's request contradict the policy, it is promptly rejected. The User/Developer is informed of this decision through an error message that highlights the specific policy violation, akin to:
+
+```
+The pods "my-pod" is invalid: : ValidatingAdmissionPolicy 'control-plane-scheduling-policy' with binding 'control-plane-scheduling-policy-binding' denied request: failed expression: object.spec.tolerations.all(toleration, !(toleration.key == 'node-role.kubernetes.io/control-plane' && toleration.effect == 'NoExecute'))
+```
+
+This message clearly indicates the failed policy expression, aiding Users/Developers in understanding the reason behind the rejection and guiding them towards necessary adjustments.
+
+- Conversely, if the request aligns with the policy or falls within an exception due to namespace labeling, the process advances eventually adhering to Kubernetes' standard scheduling steps (or direct pod placement with spec.nodeName).
+
+8. Emergency Overrides and Adjustments:
+
+- For emergency scenarios necessitating temporary deviations from the norm, Cluster Administrators have the option to swiftly modify namespace labels or adjust the ValidatingAdmissionPolicy and ValidatingAdmissionPolicyBinding resources. This flexibility ensures that critical operations can proceed unhindered, even under exceptional circumstances.
+
+**Note:** While the ValidatingAdmissionPolicy feature is in Tech Preview within OpenShift, an initial step requires Cluster Administrators to enable the TechPreview feature gate on their cluster and subsequently restart the API Servers to facilitate the creation of policies and bindings. It's important to note that activating this feature marks the cluster as non-upgradable. This significant consideration should be carefully weighed when deciding to implement this enforcement approach.
```

Comment:
Do we have a forecast of when ValidatingAdmissionPolicy will GA in OpenShift? Are there any trackers we might want to link to?

Reply:
I think it comes enabled by default with kube 1.30... Not sure we would not enable it downstream when that happens. Would have to check :thinking: